services/kvstore/src/main/java/org/hyperledger/besu/services/kvstore/LayeredKeyValueStorage.java

```diff
@@ -178,61 +178,58 @@ public Stream<Pair<byte[], byte[]>> stream(final SegmentIdentifier segmentId) {
+        new PeekingIterator<>(ourLayerState.entrySet().stream().iterator());
+    PeekingIterator<Pair<byte[], byte[]>> parentIterator =
         new PeekingIterator<>(parent.stream(segmentId).iterator());
 
     return StreamSupport.stream(
-        Spliterators.spliteratorUnknownSize(
-            new Iterator<>() {
-              @Override
-              public boolean hasNext() {
-                return ourIterator.hasNext() || parentIterator.hasNext();
-              }
-
-              private Pair<byte[], byte[]> mapEntryToPair(
-                  final Map.Entry<Bytes, Optional<byte[]>> entry) {
-                return Optional.of(entry)
-                    .map(
-                        e ->
-                            Pair.of(
-                                e.getKey().toArrayUnsafe(),
-                                e.getValue().orElseGet(() -> new byte[0])))
-                    .get();
-              }
-
-              @Override
-              public Pair<byte[], byte[]> next() {
-                var ourPeek = ourIterator.peek();
-                var parentPeek = parentIterator.peek();
-
-                if (ourPeek == null || parentPeek == null) {
-                  return ourPeek == null
-                      ? parentIterator.next()
-                      : mapEntryToPair(ourIterator.next());
-                }
-
-                // otherwise compare:
-                int comparison = ourPeek.getKey().compareTo(Bytes.wrap(parentPeek.getKey()));
-                if (comparison < 0) {
-                  return mapEntryToPair(ourIterator.next());
-                } else if (comparison == 0) {
-                  // skip dupe key from parent, return ours:
-                  parentIterator.next();
-                  return mapEntryToPair(ourIterator.next());
-                } else {
-                  return parentIterator.next();
-                }
-              }
-            },
-            ORDERED | SORTED | DISTINCT),
-        false);
+            Spliterators.spliteratorUnknownSize(
+                new LayeredIterator(ourIterator, parentIterator), ORDERED | SORTED | DISTINCT),
+            false)
+        .filter(e -> e.getValue() != null);
+  }
+
+  private static class LayeredIterator implements Iterator<Pair<byte[], byte[]>> {
+    private final PeekingIterator<Map.Entry<Bytes, Optional<byte[]>>> ourIterator;
+    private final PeekingIterator<Pair<byte[], byte[]>> parentIterator;
+
+    LayeredIterator(
+        final PeekingIterator<Map.Entry<Bytes, Optional<byte[]>>> ourIterator,
+        final PeekingIterator<Pair<byte[], byte[]>> parentIterator) {
+      this.ourIterator = ourIterator;
+      this.parentIterator = parentIterator;
+    }
+
+    @Override
+    public boolean hasNext() {
+      return ourIterator.hasNext() || parentIterator.hasNext();
+    }
+
+    private Pair<byte[], byte[]> mapEntryToPair(final Map.Entry<Bytes, Optional<byte[]>> entry) {
+      byte[] value = entry.getValue().orElse(null);
```

Comment:
why does this now return null if they value is empty instead of new byte[0]?

Reply:
Since the empty entries have already been filtered out in the current code, we won't encounter an `orElse` scenario here, right? This code should never happen. 

Since removing the initial filter, I'm setting it to null so I can remove it in the final stream.

Reply:
Thanks, that makes sense now